### PR TITLE
ARROW-7903: [Rust] [DataFusion] Migrated to sqlparser 0.6.1

### PR DESCRIPTION
**Pull request by jorgecarleitao**
AFAIK this keeps all functionality that we have available and open the doors to all the goodies of the newer version. There are too many to enumerate here, but the most predominant for me are multi statements and sub-queries.

As a side note, sqlparser is very well designed and easy-to-use, and I am convinced that we should continue to take this dependency. We currently need some custom parsing and it was very easy to understand what I needed to do based on the documentation and their code.

I am not sure this PR keeps the same SQL notation for registering CSV tables: this PR offloads the parsing of the optional schema to sqlparser, and thus if we were accepting non-compliant SQL, then we no longer accept it. I think that the parser would benefit from tests in this area.

### CODE REVIEW

**Comment by andygrove** [comment]
Thanks @jorgecarleitao this is fantastic. I imagined this would be a much larger change. I took a quick look through and LGTM but I need to find time for a more thorough review.

File: rust/datafusion/src/execution/context.rs [hunk 1]
```diff
@@ -140,10 +142,10 @@ impl ExecutionContext {
 
     /// Creates a logical plan
     pub fn create_logical_plan(&mut self, sql: &str) -> Result<LogicalPlan> {
-        let ast = DFParser::parse_sql(sql)?;
+        let statements = DFParser::parse_sql(sql)?;
 
-        match ast {
-            DFASTNode::ANSI(ansi) => {
+        match &statements[0] {
```
Comment (andygrove) [inline rust/datafusion/src/execution/context.rs:147]: I think we should return an `Err` result here if `statements.len() != 1`.

File: rust/datafusion/src/sql/parser.rs [hunk 2]
```diff
@@ -72,153 +74,185 @@ pub struct DFParser {
...
         Ok(DFParser {
             parser: Parser::new(tokens),
         })
     }
 
-    /// Parse a SQL statement and produce an Abstract Syntax Tree (AST)
-    pub fn parse_sql(sql: &str) -> Result<DFASTNode, ParserError> {
+    /// Parse a SQL statement and produce a set of statements
+    pub fn parse_sql(sql: &str) -> Result<Vec<Statement>, ParserError> {
+        let mut tokenizer = Tokenizer::new(&GenericDialect {}, &sql);
```
Comment (andygrove) [inline rust/datafusion/src/sql/parser.rs:87]: Just a side note but I think eventually we should use the Hive dialect.

File: rust/datafusion/src/sql/parser.rs [hunk 3]
```diff
@@ -72,153 +74,185 @@ pub struct DFParser {
...
+    /// Report unexpected token
+    fn expected<T>(&self, expected: &str, found: Token) -> Result<T, ParserError> {
+        parser_err!(format!("Expected {}, found: {}", expected, found))
     }
 
     /// Parse a new expression
-    pub fn parse(&mut self) -> Result<DFASTNode, ParserError> {
-        self.parse_expr(0)
+    pub fn parse_statement(&mut self) -> Result<Statement, ParserError> {
+        match self.parser.next_token() {
```
Comment (andygrove) [inline rust/datafusion/src/sql/parser.rs:119]: Perhaps it would be better to use `peek_token` here to avoid the need to roll back?

**Review by andygrove — APPROVED** [review]
I left some comments but LGTM overall.

**Comment by andygrove** [comment]
@houqp fyi

File: rust/datafusion/src/execution/context.rs (hunk 1 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/execution/context.rs:153]: Well spotted. Fixed

File: rust/datafusion/src/sql/parser.rs (hunk 3 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/sql/parser.rs]: done.

**Review by andygrove — APPROVED** [review]
LGTM

**Review by houqp — APPROVED** [review]

**Pull request closed without merging**